### PR TITLE
Graph-Kiota software requirements specifications

### PR DESCRIPTION
**Pull request by nikithauc**
Creating this PR to open a discussion and capture the specifications as we go along with the implementation of Graph and Kiota.

There are `Kiota` files here which we can move after we have a concrete set of requirements and implementations. Till then let's maintain one place to capture the decisions we make.

### CODE REVIEW

File: graph-kiota/graph-kiota.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+### Graph OpenAPI specification
+
+- OpenAPI v1 specification:https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/v1.0/openapi.yaml
+- OpenAPI beta specification:https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/beta/openapi.yaml
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:4]:
```suggestion
- [OpenAPI beta specification](https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/beta/openapi.yaml)
```

File: graph-kiota/graph-kiota.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+### Graph libraries using Kiota
+
+#### Graph Service library:
+
+The Graph Service libraries which contain the request builders will be generated using Kiota
+
+Generation command:
+```
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:13]:
```suggestion
```Shell
```

File: graph-kiota/graph-kiota.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+kiota.exe -d https://raw.githubusercontent.com/microsoftgraph/msgraph-metadata/master/openapi/v1.0/openapi.yaml --language typescript -o path\msgraph-sdk-typescript\src -c GraphServiceClient -n MicrosoftGraph
+```
+- GraphServiceClient generated.
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:16]: Can I suggest adding a link to the parameters reference? https://github.com/microsoft/kiota/#parameters-reference

File: graph-kiota/graph-kiota.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    - Core client with the user configurations such as authprovider, custom middleware, request options.
+    - HttpCore instance for the service library with user configurations such as authprovider, custom middleware, request options.
+
+- Tasks
+    - PageIteratorTask
+    - LargeFileUploadTask
+    - Batching
+
+- Authentication Providers
+    - Use the `IAuthProvider` and  auth provider class from the Kiota Auth abstractions.
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:63]:
```suggestion
    - Use the `IAuthenticationProvider` and  authentication provider class from the Kiota abstractions.
```

File: graph-kiota/graph-kiota.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+
+#### Graph Service library specifications
+
+- Discuss about ways to add custom code or custom configurations to the generated `GraphServiceClient` code.
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:68]: I'm not sure I get what you mean by this statement, can you expand?

File: graph-kiota/graph-kiota.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+- The custom configurations that can go here are as follows: 
+ - Setting the default for `HttpCore` with a Graph specific `HttpCore` instance. 
+ - Adding multiple constructors to the `GraphServiceClient`. 
+
+#### Graph Middlewares
+
+- Import middlewares from Kiota core library
+- How to add graph specific checks in the middleware. Example: The JS SDK checks if the url is a Graph url or a custom host passed by the user before appending telemetry headers or auth headers. If not, the headers are removed.
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:76]: the generic middleware could accept a set of valid hosts for example. @andrueastman did it for dotnet so the telemetry handler accepts a callback, effectively giving full control to graph core here. Some alignment work required across languages for sure.

File: graph-kiota/graph-kiota.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+- AuthHandler - Kiota core
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:78]:
```suggestion
- AuthHandler - removed
```
This is now handled by the auth providers:
- This avoids having to configure both the middleware and the provider
- The authentication provider implementations are available across any http client

File: graph-kiota/graph-kiota.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+#### Use cases
+
+- Graph core client 
+```
+var response = new coreClient.requestURL("/me").get();
+```
+
+- Graph service client
+```
+var response = new serviceClient.me().get();
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:93]:
```suggestion
var response = await serviceClient.me.get();
```

I believe we're generating getters for typescript for navigation like that. I'd have to double check

File: graph-kiota/graph-kiota.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+- AuthHandler - Kiota core
+- RetryHandler - Kiota core
+- RedirectHandler - Kiota core
+- CompressionHandler - Kiota core
+- CachingHandler - Kiota core
+- ChaosHandler - Graph core 
+#### Use cases
+
+- Graph core client 
+```
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:87]:
```suggestion
```TypeScript
```

(for all the snippets below)

File: graph-kiota/graph-kiota.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+- The service library should extend the capabilities of the core library, that is, a user using service client should not require to create or use a core client to access any functionality.
+-  Some options to  achieve this (TBD):
+ -  Maintain a parent abstract class for the core and the service client.
+ -  `GraphServiceClient` extends the `GraphCoreClient`.
+
+### More questions
+
+- Does the generated ApiClient have to be tightly coupled with the generated request builders? Should it always be tightly coupled? [MEM] I ask as there could be scenarios where the request builders are not necessary and the client developers just wants the Kiota.Core functionality in the client, and will use client.requestUrl<Response>.Get();
```
Comment (baywet) [inline graph-kiota/graph-kiota.md:125]: the api client is only a request builder with a special constructor and fewer properties. So in that sense yes. This was an early design choice.

File: graph-kiota/httpCore.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
+### IHttpCore
```
Comment (baywet) [inline graph-kiota/httpCore.md:1]: this is all already documented [here](https://github.com/microsoft/kiota/blob/main/docs/kiotaabstractions.md#http-core) and I'd much rather have that information at this place so it's easier to find for people outside of Microsoft. Don't hesitate to go PR it.

File: graph-kiota/kiota.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+Learn [Kiota](https://github.com/microsoft/kiota#readme)
+
+### APIs that can use Kiota 
+- Any HTTP API with an Open API specification (For generators). Not limited to Microsoft Graph.
+
+### Kiota libraries
+
+Common for all languagues:
+- Kiota generator: Generates the request builders
+The generated code contains the APIClient class and the request builders.
```
Comment (baywet) [inline graph-kiota/kiota.md:11]: I'd rephrase to request builders for fluent API and models. Api Client is just a special request builder.

File: graph-kiota/kiota.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    httpCore : HttpCore;
+
+    getResourceById(): ResourceRequestBuilder;
+}
+```
+
+Libraries for each language:
+
+- Kiota abstractions
+    - This library contains only abstract classes and interfaces which can be implemented to connect the generated request builders with a client library that makes the call to the API.
```
Comment (baywet) [inline graph-kiota/kiota.md:24]:
```suggestion
    - This library contains only abstract classes and interfaces which can be implemented to connect the generated request builders with a client library that makes the call to the API, handles serialization and authentication.
```

File: graph-kiota/kiota.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    [Kiota abstractions dotnet](https://github.com/microsoft/kiota/tree/main/abstractions/dotnet)
+    [Kiota abstractions java](https://github.com/microsoft/kiota/tree/main/abstractions/java)
+    [Kiota abstractions typescript](https://github.com/microsoft/kiota/tree/main/abstractions/typescript)
+
+    - Contains the following:
+        - HttpCore interface
+
+        ```
```
Comment (baywet) [inline graph-kiota/kiota.md:32]: again, let's refer to https://github.com/microsoft/kiota/blob/main/docs/kiotaabstractions.md#http-core instead of duplicating this.

File: graph-kiota/kiota.md [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+   
+            sendAsync<ModelType extends Parsable>(requestInfo: RequestInformation, type: new() => ModelType, responseHandler: ResponseHandler | undefined): Promise<ModelType>;
+            sendCollectionAsync<ModelType extends Parsable>(requestInfo: RequestInformation, type: new() => ModelType, responseHandler: ResponseHandler | undefined): Promise<ModelType[]>;
+            sendPrimitiveAsync<ResponseType>(requestInfo: RequestInformation, responseType: "string" | "number" | "boolean" | "Date" | "ReadableStream", responseHandler: ResponseHandler | undefined): Promise<ResponseType>;
+            sendNoResponseContentAsync(requestInfo: RequestInformation, responseHandler: ResponseHandler | undefined): Promise<void>;
+            enableBackingStore(backingStoreFactory?: BackingStoreFactory | undefined): void;
+        }
+        ````
+        - Serialization interfaces
+        - Authentication: `IAuthProvider`, abstract classes- `anonymousAuthenticationProvider` and `BearerTokenAuthneticationProvider`
```
Comment (baywet) [inline graph-kiota/kiota.md:44]:
```suggestion
        - Authentication: `IAuthenticationProvider`, abstract classes- `anonymousAuthenticationProvider` and `BearerTokenAuthneticationProvider`
```

File: graph-kiota/kiota.md [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        - Utils
+        - ResponseHandler interface
+        - MiddlewareOption class
```
Comment (baywet) [inline graph-kiota/kiota.md:47]:
```suggestion
        - RequestOption interface
```
Anticipating the rename

File: graph-kiota/kiota.md [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+
+- Kiota core
+    - This is generic core client library (functionality or purpose similar to Microsoft Graph core libraries)
+    [Kiota core dotnet](https://github.com/microsoft/kiota/tree/main/http/dotnet/httpclient)
+    [Kiota core java](https://github.com/microsoft/kiota/tree/main/http/java/okhttp)
+    [Kiota core typescript](https://github.com/microsoft/kiota/tree/main/http/typescript/fetch)
+    - This library contains the following:
+        -  HttpCore class implementing the abstraction HttpCore interface.
+        -  HttpClientbuilder (.NET, Java) or HTTPClient class (TS) (processes and records the middleware chain used while making requests.)
+        -  Middlewares: AuthHandler, Retry, Redirect, Compression, Caching, Telemetry
```
Comment (baywet) [inline graph-kiota/kiota.md:59]:
```suggestion
        -  Middlewares: RetryHandler, Redirect, Compression, Caching, Telemetry
```

File: graph-kiota/kiota.md [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+    
+```
+class HttpCore implements kiota.abstractions.HttpCore {
```
Comment (baywet) [inline graph-kiota/kiota.md:62]: let's leave that on the Kiota repo and refer to it

File: graph-kiota/kiota.md [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+    SampleRequestBuilder(httpCoreInstance);
+    ExampleRequestBuilder(httpCoreInstance);
+}
+```
+
+- Kiota serializations 
+    - This library contains implementation of the kiota abstractions serialization interfaces
+    - Currently targets `JSON` format.
+    - Other formats to consider:
+        - Multipart payload
```
Comment (baywet) [inline graph-kiota/kiota.md:94]: we did explore the multi-part aspect with @andrueastman and decided to leave it for now. Only OneNote uses it to our knowledge.

File: graph-kiota/kiota.md [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+        - Base64 encoded string
```
Comment (baywet) [inline graph-kiota/kiota.md:95]: the case where it's part of a json payload should be handled today already. If the payload is only base64 encoded, this needs to be reviewed.

File: graph-kiota/kiota.md [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+        - text/plain
+        - xml
+
+- Kiota auth providers
+
+### Example of how we use the Kiota libraries:
+
+1. Clone the Kiota repo, build the project and navigate the path containing the `kiota.exe`
```
Comment (baywet) [inline graph-kiota/kiota.md:103]: can we refer to this documentation, and improve it if required? https://github.com/microsoft/kiota/#generating-sdks

File: graph-kiota/kiota.md [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+2. Generate the request builder using the command:
+```
+kiota.exe -d ../msgraph-sdk-powershell/openApiDocs/v1.0/mail.yml --language csharp -o ../somepath -n namespaceprefix
+```
+3. The generation should contain the `APIClient` class and request builders.
+
+5. Create a demo class to test the `APIClient`.
+
+4. Import the `KiotaCore` library
+- [dotnet](https://github.com/microsoft/kiota/packages/826853)
```
Comment (baywet) [inline graph-kiota/kiota.md:113]:
```suggestion
- [dotnet](https://github.com/microsoft/kiota/blob/main/docs/requiredtools/dotnet.md)
```

File: graph-kiota/kiota.md [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+- [TypeScript](https://github.com/microsoft/kiota/packages/827040)
```
Comment (baywet) [inline graph-kiota/kiota.md:114]:
```suggestion
- [TypeScript](https://github.com/microsoft/kiota/blob/main/docs/requiredtools/typescript.md)
```

File: graph-kiota/kiota.md [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+
+```
+import {JsonParseNodeFactory, JsonSerializationWriterFactory} from '@microsoft/kiota-serialization-json';
+```
+
+Example:
+
+6. Import the AuthProvider library or implement the IAuthProvider interface to create a custom AuthProvider class.
+
+Find a suitable Kiota Authentication library in https://github.com/microsoft/kiota/packages
```
Comment (baywet) [inline graph-kiota/kiota.md:133]:
```suggestion
Find a suitable Kiota Authentication library in https://github.com/microsoft/kiota/tree/main/authentication/typescript/azure
```

We could add a rollup of all the authentication packages in the /authentication folder, and one by language in the /authentication/language folder if it makes things easier.

File: graph-kiota/kiota.md [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+
+```
+var httpCore = new HTTPCore(authProvider, JsonParseNodeFactory, JsonSerializationWriterFactory)
+```
+
+8. Create a APIClient instance:
+
+```
+var apiClient = new APIClient(httpCore);
+var response = apiClient.users("<ID>").get();
```
Comment (baywet) [inline graph-kiota/kiota.md:146]:
```suggestion
var response = await apiClient.users("<ID>").get();
```

File: graph-kiota/graph-kiota.md [hunk 26]
```diff
@@ -0,0 +1,126 @@
+### Graph OpenAPI specification
```
Comment (darrelmiller) [inline graph-kiota/graph-kiota.md:1]:
```suggestion
### Graph OpenAPI description
```
The specification is the document that describes OpenAPI syntax.  API descriptions are what are created to describe APIs using the OpenAPI specification.

File: graph-kiota/kiota.md (hunk 21 above)
Comment (darrelmiller) [reply graph-kiota/kiota.md:103]: There are language specific pages here https://github.com/microsoft/kiota/tree/main/docs/requiredtools that show how to use the Kiota tooling.  There dotnet one has an end to end experience.  It would be great to do the same for JS/TS

File: graph-kiota/kiota.md (hunk 14 above)
Comment (nikithauc) [reply graph-kiota/kiota.md:32]: Created this here to capture the description and use case of each send function? 

Can you please provide the use cases for each of the send request, especially for `sendNoResponseContentAsync`? 

We can merge the documentations after we have captured information. 

We should update the code documentation as well.

Comment (baywet) [reply graph-kiota/kiota.md:32]: https://github.com/microsoft/kiota/blob/afc5e21da5c52a0c7fc683c4bb7c91825cdfffcd/abstractions/typescript/src/httpCore.ts#L41

File: graph-kiota/kiota.md [hunk 27]
```diff
@@ -0,0 +1,180 @@
...
+```
+
+Example:
+
+6. Import the AuthProvider library or implement the IAuthProvider interface to create a custom AuthProvider class.
+
+Find a suitable Kiota Authentication library in https://github.com/microsoft/kiota/tree/main/authentication/typescript/azure
+
+
+7. Create a HTTPCore instance:
```
Comment (baywet) [inline graph-kiota/kiota.md:136]:
```suggestion
7. Create a RequestAdapter instance:
```

File: graph-kiota/kiota.md [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+
+```
+var httpCore = new HTTPCore(authProvider, JsonParseNodeFactory, JsonSerializationWriterFactory)
```
Comment (baywet) [inline graph-kiota/kiota.md:139]:
```suggestion
var requestAdapter = new FetchRequestAdapter(authProvider, JsonParseNodeFactory, JsonSerializationWriterFactory)
```

File: graph-kiota/kiota.md [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+```
+
+8. Create a APIClient instance:
+
+```
+var apiClient = new APIClient(httpCore);
```
Comment (baywet) [inline graph-kiota/kiota.md:145]:
```suggestion
var apiClient = new APIClient(requestAdapter);
```

**Review by baywet — CHANGES_REQUESTED** [review]
